src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -70,7 +70,72 @@ public sealed partial class JsonSerializerOptions
             return converters;
 
             void Add(JsonConverter converter) =>
-                converters.Add(converter.TypeToConvert!, converter);
+                converters.Add(converter.TypeToConvert, converter);
+        }
+
+        internal JsonConverter GetDictionaryKeyConverter(Type keyType)
+        {
+            if (_dictionaryKeyConverters == null)
+            {
+                _dictionaryKeyConverters = GetDictionaryKeyConverters();
+            }
+
+            if (!_dictionaryKeyConverters.TryGetValue(keyType, out JsonConverter? converter)){
+                if (keyType.IsEnum)
+                {
+                    converter = GetEnumConverter();
+                    _dictionaryKeyConverters[keyType] = converter;
+                }
+                else
+                {
+                    ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(keyType);
+                }
+            }
+
+            return converter!;
+
+            JsonConverter GetEnumConverter()
+                => (JsonConverter)Activator.CreateInstance(
+                        typeof(EnumConverter<>).MakeGenericType(keyType),
+                        BindingFlags.Instance | BindingFlags.Public,
+                        binder: null,
+                        new object[] { EnumConverterOptions.AllowStrings | EnumConverterOptions.AllowNumbers, this },
+                        culture: null)!;
```

Comment:
Should this code live in the `EnumConverterFactory`?

Reply:
I think is better to have it here since it is only useful in this method. 
I can add a comment explaining that this is performed to support multiple `enum` types as dictionary key and we want to use our own defined semantics (we don't want to call a custom converter).